File: 1458-max-dot-product-of-two-subsequences/1458-max-dot-product-of-two-subsequences.cpp
class Solution {
public:
       int maxDotProduct(vector<int>& nums1, vector<int>& nums2) {
        int m = nums1.size();
        int n = nums2.size();
     
        vector<vector<int>> dp(m + 1, vector<int>(n + 1, INT_MIN));
        
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                int product = nums1[i-1] * nums2[j-1];
                int withPrev = (dp[i-1][j-1] == INT_MIN) ? product : dp[i-1][j-1] + product;
               
                int justCurrent = product;
                
                
                int skipFirst = dp[i-1][j];
                
                
                int skipSecond = dp[i][j-1];
                
                dp[i][j] = max({withPrev, justCurrent, skipFirst, skipSecond});
            }
        }
        
        return dp[m][n];
    }
};

class SolutionOptimized {
public:
    int maxDotProduct(vector<int>& nums1, vector<int>& nums2) {
        int m = nums1.size();
        int n = nums2.size();
        
        vector<int> prev(n + 1, INT_MIN);
        vector<int> curr(n + 1, INT_MIN);
        
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                int product = nums1[i-1] * nums2[j-1];
                
                int withPrev = (prev[j-1] == INT_MIN) ? product : prev[j-1] + product;
                int justCurrent = product;
                int skipFirst = prev[j];
                int skipSecond = curr[j-1];
                
                curr[j] = max({withPrev, justCurrent, skipFirst, skipSecond});
            }
            swap(prev, curr);
            fill(curr.begin(), curr.end(), INT_MIN);
        }
        
        return prev[n];
    }
};

File: 1458-max-dot-product-of-two-subsequences/1458-max-dot-product-of-two-subsequences.rs
impl Solution {
    pub fn max_dot_product(nums1: Vec<i32>, nums2: Vec<i32>) -> i32 {
        let m = nums1.len();
        let n = nums2.len();
        
        let mut dp = vec![vec![i32::MIN; n + 1]; m + 1];
        
        for i in 1..=m {
            for j in 1..=n {
                let product = nums1[i-1] * nums2[j-1];
                
                let with_prev = if dp[i-1][j-1] == i32::MIN {
                    product
                } else {
                    dp[i-1][j-1] + product
                };
                let just_current = product;
                
                let skip_first = dp[i-1][j];
                
                let skip_second = dp[i][j-1];
                
                dp[i][j] = with_prev.max(just_current).max(skip_first).max(skip_second);
            }
        }
        
        dp[m][n]
    }
}
